### PR TITLE
[Platform] Speedups

### PR DESCRIPTION
**Pull request by benaadams**
Speed up the platform application
* Move sync request processing out of `async ProcessRequest` and into `ParseHttpRequest`, also saves passing large structs. https://github.com/aspnet/KestrelHttpServer/pull/2569/commits/b33f673d6c70603f3817fffeda39ada7fe226890
* Use struct adaptor (as per Kestrel) for parsing to convert the interface calls to virtual calls 
https://github.com/aspnet/KestrelHttpServer/pull/2569/commits/baab995311a680aefa6465c427cbcc55880d7e8a and https://github.com/aspnet/KestrelHttpServer/pull/2569/commits/ae8c0d04c0d78a1ee74eb46a32419b3b650737da
    - `[FAILED: target not direct] IHttpHeadersHandler:OnHeader(struct,struct):this` <-- changes
    - `[below ALWAYS_INLINE size] ParsingAdapter:OnHeader(struct,struct):this +
       [FAILED: target not direct] HttpConnection:OnHeader(struct,struct):this` <-- to
* Remove the virtual calls to make them direct calls; which also causes the `OnHeader` call to inline and elide https://github.com/aspnet/KestrelHttpServer/pull/2569/commits/56182286015c9031769478d196990cb1f07ef1e5
    - `[FAILED: target not direct] HttpConnection:OnHeader(struct,struct):this` <-- changes
    - `[below ALWAYS_INLINE size] BenchmarkApplication:OnHeader(struct,struct):this` <-- to
* Request type to `enum` rather than double `bool` https://github.com/aspnet/KestrelHttpServer/pull/2569/commits/044f1bf57030b5a58369da2b3cb7567ed24bf439
* Pre-init response statics to avoid `CORINFO_HELP_GETSHARED_NONGCSTATIC_BASE` per call https://github.com/aspnet/KestrelHttpServer/pull/2569/commits/efdee91414a7e353816006944f2f0f3cdb295dbf
* Pre-init path statics to avoid `CORINFO_HELP_GETSHARED_NONGCSTATIC_BASE` per call https://github.com/aspnet/KestrelHttpServer/pull/2569/commits/cf910e35d902b8265ff4c6dddca647b912a82250
* Faster Json fall through: Pre-check path lengths prior to calling into `SpanHelpers:SequenceEqual` for match https://github.com/aspnet/KestrelHttpServer/pull/2569/commits/0704d02cac831d10cf7c8f1a861a052eea664170
* Use `CountingBufferWriter<T>` as non-shared generic (via struct wrapper) https://github.com/aspnet/KestrelHttpServer/pull/2569/commits/ff3fdeb66e7d4e65ed27ae85e6f9ffc0d36e22ea
    -  `call     qword ptr [rax]IBufferWriter'1:Advance(int):this` <-- changes
    -  `call     qword ptr [rax]PipeWriter:Advance(int):this` <-- to
* Overwrite buffer if slice required, rather than than using slice as part of ternary 
https://github.com/aspnet/KestrelHttpServer/pull/2569/commits/58d2ee32a6584f9e4289b8f86b2a89b8ca04a859 https://github.com/aspnet/KestrelHttpServer/pull/2569#issuecomment-388655781
* Pre-init date static 
https://github.com/aspnet/KestrelHttpServer/pull/2569/commits/fdea4abd6cd8e30a6c2f54f6fe746ab7d23439f7 https://github.com/aspnet/KestrelHttpServer/pull/2569#issuecomment-388659044
* Smaller WriteNumeric method 
https://github.com/aspnet/KestrelHttpServer/pull/2569/commits/71c9d3d4eb9a799bbe4d8ebd96be78c4ccc56389 https://github.com/aspnet/KestrelHttpServer/pull/2569#issuecomment-388662068
* Store parsing state transition in local (register) rather than class (memory) 
https://github.com/aspnet/KestrelHttpServer/pull/2569/commits/e450af6a25419d843b7c84db55f2e9683f9d2d8c https://github.com/aspnet/KestrelHttpServer/pull/2569#issuecomment-388664482
* CountingBufferWriter -> smaller BufferWriter (drop unused long field) 
https://github.com/aspnet/KestrelHttpServer/pull/2569/commits/d9b3c3c6a8308df3cd35a87dfc4c7ee1bfe0e72e https://github.com/aspnet/KestrelHttpServer/pull/2569#issuecomment-388666609

### CODE REVIEW

**Comment by benaadams** [comment]
Hoping for more than +3% 

image

https://www.techempower.com/benchmarks/#section=test&runid=415d1b9b-883d-4c41-bea5-82fe685a97d4&hw=ph&test=plaintext

**Comment by benaadams** [comment]
Pre
```
[FAILED: target not direct] IHttpHeadersHandler:OnHeader(struct,struct):this

; Total bytes of code 7081, prolog size 73 for method 
 HttpParser`1:ParseHeaders(ref,byref,byref,byref,byref):bool:this
```
Post
```
[below ALWAYS_INLINE size] ParsingAdapter:OnHeader(struct,struct):this
 [below ALWAYS_INLINE size] BenchmarkApplication:OnHeader(struct,struct):this

; Total bytes of code 6956, prolog size 73 for method
 HttpParser`1:ParseHeaders(struct,byref,byref,byref,byref):bool:this
```

**Comment by benaadams** [comment]
K, think am done - updated summary

**Comment by sebastienros** [comment]
It's super close to the noise level, but I could measure at least 2% of gain on Citrine. A little more the old physical.

File: benchmarkapps/PlatformBenchmarks/PipelineExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+using System.Runtime.InteropServices;
+
+namespace PlatformBenchmarks
+{
+    // Same as KestrelHttpServer\src\Kestrel.Core\Internal\Http\PipelineExtensions.cs
+    // However methods accept T : struct, IBufferWriter<byte> rather than PipeWriter.
+    // This allows a struct wrapper to turn CountingBufferWriter into a non-shared generic,
+    // while still offering the WriteNumeric extension.
+
+    public static class PipelineExtensions
```
Comment (davidfowl) [inline benchmarkapps/PlatformBenchmarks/PipelineExtensions.cs]: Rename this to BufferExtensions

Comment (benaadams) [reply benchmarkapps/PlatformBenchmarks/PipelineExtensions.cs]: Done

**Comment by benaadams** [comment]
Reuse var for slice
```diff
 ; Assembly listing for method BenchmarkApplication:ParseHttpRequest(byref):bool:this
 ...
-; Lcl frame size = 272
+; Lcl frame size = 208
 ...

-       4881EC10010000       sub      rsp, 272
+       4881ECD0000000       sub      rsp, 208
        C5F877               vzeroupper 
        488BF1               mov      rsi, rcx
        488D7C2438           lea      rdi, [rsp+38H]
-       B936000000           mov      ecx, 54
+       B926000000           mov      ecx, 38
        33C0                 xor      rax, rax
        F3AB                 rep stosd 
...  
- G_M38489_IG10:
-        C4E17A6F8424F0000000 vmovdqu  xmm0, qword ptr [rsp+F0H]
-        C4E17A7F842488000000 vmovdqu  qword ptr [rsp+88H], xmm0
-        C4E17A6F842400010000 vmovdqu  xmm0, qword ptr [rsp+100H]
-        C4E17A7F842498000000 vmovdqu  qword ptr [rsp+98H], xmm0
...
- G_M38489_IG16:
-        C4E17A6F842488000000 vmovdqu  xmm0, qword ptr [rsp+88H]
-        C4E17A7F8424A8000000 vmovdqu  qword ptr [rsp+A8H], xmm0
-        C4E17A6F842498000000 vmovdqu  xmm0, qword ptr [rsp+98H]
-        C4E17A7F8424B8000000 vmovdqu  qword ptr [rsp+B8H], xmm0
...
-; Total bytes of code 874, prolog size 40 for method BenchmarkApplication:ParseHttpRequest(byref):bool:this
+; Total bytes of code 787, prolog size 40 for method BenchmarkApplication:ParseHttpRequest(byref):bool:this
```

**Comment by benaadams** [comment]
Pre-init date static
```diff
 ; Assembly listing for method BenchmarkApplication:PlainText(ref)
 ...
 ; Lcl frame size = 248
 ...
 G_M36647_IG20:
        48B94829EC53A1020000 mov      rcx, 0x2A153EC2948
-       BA02000000           mov      edx, 2
-       E8D561825F           call     CORINFO_HELP_GETSHARED_NONGCSTATIC_BASE
-       48B94829EA43F7010000 mov      rcx, 0x1F743EA2948
        488B09               mov      rcx, gword ptr [rcx]
        4885C9               test     rcx, rcx
        7506                 jne      SHORT G_M36647_IG21
        33D2                 xor      rdx, rdx
        33C0                 xor      eax, eax
        EB07                 jmp      SHORT G_M36647_IG22
...
-; Total bytes of code 2238, prolog size 32 for method BenchmarkApplication:PlainText(ref)
+; Total bytes of code 2218, prolog size 32 for method BenchmarkApplication:PlainText(ref)
```

**Comment by benaadams** [comment]
Smaller WriteNumeric
```diff
 ; Assembly listing for method BenchmarkApplication:PlainText(ref)
 ...
-       4863D1               movsxd   rdx, ecx
...
-       41B801000000         mov      r8d, 1
+       4533C0               xor      r8d, r8d
...
-       8808                 mov      byte  ptr [rax], cl
-       8B8424D8000000       mov      eax, dword ptr [rsp+D8H]
-       FFC0                 inc      eax
-       898424D8000000       mov      dword ptr [rsp+D8H], eax
-       83BC24F000000001     cmp      dword ptr [rsp+F0H], 1
-       0F826A030000         jb       G_M36647_IG71
-       488D8C24E8000000     lea      rcx, bword ptr [rsp+E8H]
-       488B09               mov      rcx, bword ptr [rcx]
-       8B8424F0000000       mov      eax, dword ptr [rsp+F0H]
-       FFC8                 dec      eax
-       48FFC1               inc      rcx
-       4C8D8C24E8000000     lea      r9, bword ptr [rsp+E8H]
-       498909               mov      bword ptr [r9], rcx
-       41894108             mov      dword ptr [r9+8], eax
-       E93A010000           jmp      G_M36647_IG48 
+       448800               mov      byte  ptr [rax], r8b
+       41B801000000         mov      r8d, 1
+       E99C000000           jmp      G_M36637_IG44
        4883FA64             cmp      rdx, 100
-       0F837F000000         jae      G_M36647_IG45
+       7339                 jae      SHORT G_M36637_IG43
        83F902               cmp      ecx, 2
-       7C7A                 jl       SHORT G_M36647_IG45
-       8BCA                 mov      ecx, edx
+       7C34                 jl       SHORT G_M36637_IG43
...	   
-       8B8424D8000000       mov      eax, dword ptr [rsp+D8H]
-       83C002               add      eax, 2
-       898424D8000000       mov      dword ptr [rsp+D8H], eax
-       83BC24F000000002     cmp      dword ptr [rsp+F0H], 2
-       0F82E8020000         jb       G_M36647_IG72
+       41B802000000         mov      r8d, 2
+       EB5E                 jmp      SHORT G_M36637_IG44
-       488D8C24E8000000     lea      rcx, bword ptr [rsp+E8H]
-       488B09               mov      rcx, bword ptr [rcx]
-       8B8424F0000000       mov      eax, dword ptr [rsp+F0H]
-       83C0FE               add      eax, -2
-       4883C102             add      rcx, 2
-       4C8D8C24E8000000     lea      r9, bword ptr [rsp+E8H]
-       498909               mov      bword ptr [r9], rcx
-       41894108             mov      dword ptr [r9+8], eax
-       E9B1000000           jmp      G_M36647_IG48
...
-       0F83A1000000         jae      G_M36647_IG47
+       7356                 jae      SHORT G_M36637_IG44
       83F903               cmp      ecx, 3
-       0F8C98000000         jl       G_M36647_IG47
+       7C51                 jl       SHORT G_M36637_IG44
-       8BCA                 mov      ecx, edx
...
-       8B8C24D8000000       mov      ecx, dword ptr [rsp+D8H]
-       83C103               add      ecx, 3
-       898C24D8000000       mov      dword ptr [rsp+D8H], ecx
-       83BC24F000000003     cmp      dword ptr [rsp+F0H], 3
-       0F823C020000         jb       G_M36647_IG73
+       41B803000000         mov      r8d, 3
...	   
+       48894C2460           mov      bword ptr [rsp+60H], rcx
+       4585C0               test     r8d, r8d
+       7E4A                 jle      SHORT G_M36637_IG46
+       418BC8               mov      ecx, r8d
+       038C24D8000000       add      ecx, dword ptr [rsp+D8H]
+       898C24D8000000       mov      dword ptr [rsp+D8H], ecx
+       443B8424F0000000     cmp      r8d, dword ptr [rsp+F0H]
+       0F8724020000         ja       G_M36637_IG69
...
-       33C9                 xor      rcx, rcx
-       48894C2460           mov      bword ptr [rsp+60H], rcx
-       4585C0               test     r8d, r8d
-       750D                 jne      SHORT G_M36647_IG49
       488D8C24D8000000     lea      rcx, bword ptr [rsp+D8H]
       E834D1FFFF           call     BufferExtensions:WriteNumericMultiWrite(byref,long)
...
-; Total bytes of code 2218, prolog size 32 for method BenchmarkApplication:PlainText(ref)
+; Total bytes of code 2057, prolog size 32 for method BenchmarkApplication:PlainText(ref)
```

**Comment by benaadams** [comment]
Store state in local (register)
```diff
; Assembly listing for method BenchmarkApplication:ParseHttpRequest(byref):bool:this
- ; Lcl frame size = 208
+ ; Lcl frame size = 200
...
+        4155                 push     r13
...
-        4881ECD0000000       sub      rsp, 208
+        4881ECC8000000       sub      rsp, 200
...
-        488D7C2438           lea      rdi, [rsp+38H]
+        488D7C2430           lea      rdi, [rsp+30H]
...
-        837E2400             cmp      dword ptr [rsi+36], 0
+        85ED                 test     ebp, ebp
...
-        C7462401000000       mov      dword ptr [rsi+36], 1
+        BB01000000           mov      ebx, 1
...
-        837E2401             cmp      dword ptr [rsi+36], 1
+        83FB01               cmp      ebx, 1
...
-        C7462402000000       mov      dword ptr [rsi+36], 2
+        BB02000000           mov      ebx, 2
...
-        837E2402             cmp      dword ptr [rsi+36], 2
+        83FB02               cmp      ebx, 2
...
+        895E24               mov      dword ptr [rsi+36], ebx
...
-        4881C4D0000000       add      rsp, 208
+        4881C4C8000000       add      rsp, 200
+        415D                 pop      r13
...
-        4881C4D0000000       add      rsp, 208
+        4881C4C8000000       add      rsp, 200
+        415D                 pop      r13
```

**Comment by benaadams** [comment]
Can't see anything else atm

**Comment by benaadams** [comment]
For completeness; since I started, the other commit https://github.com/aspnet/KestrelHttpServer/pull/2569/commits/d9b3c3c6a8308df3cd35a87dfc4c7ee1bfe0e72e

CountingBufferWriter -> Smaller BufferWriter
```diff
 ; Assembly listing for method BenchmarkApplication:PlainText(ref)
...
+; Lcl frame size = 264
-; Lcl frame size = 248
...
-       4881EC08010000       sub      rsp, 264
+       4881ECF8000000       sub      rsp, 248
        C5F877               vzeroupper 
        488BF1               mov      rsi, rcx
        488D7C2420           lea      rdi, [rsp+20H]
-       B93A000000           mov      ecx, 58
+       B936000000           mov      ecx, 54
        33C0                 xor      rax, rax
        F3AB                 rep stosd 
        488BCE               mov      rcx, rsi
...
-       488BD1               mov      rdx, rcx
-       33C9                 xor      rcx, rcx
...
-       48894820             mov      qword ptr [rax+32], rcx
...
-; Not inlined
-       488D8C24B8000000     lea      rcx, bword ptr [rsp+B8H]
-       E802D7FFFF           call     CountingBufferWriter`1:.ctor(struct):this
+;Inlined
+       33D2                 xor      edx, edx
+       899424B8000000       mov      dword ptr [rsp+B8H], edx
+       488D9424C0000000     lea      rdx, bword ptr [rsp+C0H]
+       48890A               mov      gword ptr [rdx], rcx
+       488D9424C8000000     lea      rdx, bword ptr [rsp+C8H]
+       4533C0               xor      r8d, r8d
+       488B01               mov      rax, qword ptr [rcx]
+       488B4048             mov      rax, qword ptr [rax+72]
+       FF5010               call     qword ptr [rax+16]PipeWriter:GetSpan(int):struct:this
...
-       488B8C24D8000000     mov      rcx, qword ptr [rsp+D8H]
-       48898C2400010000     mov      qword ptr [rsp+100H], rcx
...
-; Total bytes of code 2254, prolog size 32 for method BenchmarkApplication:PlainText(ref)
+; Total bytes of code 2238, prolog size 32 for method BenchmarkApplication:PlainText(ref)
```

**Comment by benaadams** [comment]
Can't see anything else atm

File: benchmarkapps/PlatformBenchmarks/HttpApplication.cs [hunk 2]
```diff
@@ -157,5 +150,40 @@ private enum State
...
+
+            public Memory<byte> GetMemory(int sizeHint = 0)
+                => Writer.GetMemory(sizeHint);
+
+            public Span<byte> GetSpan(int sizeHint = 0)
+                => Writer.GetSpan(sizeHint);
+        }
+    }
+
+    public struct ParsingAdapter : IHttpRequestLineHandler, IHttpHeadersHandler
```
Comment (davidfowl) [inline benchmarkapps/PlatformBenchmarks/HttpApplication.cs]: Why is this public?

Comment (benaadams) [reply benchmarkapps/PlatformBenchmarks/HttpApplication.cs]: Can be private and moved into class

File: benchmarkapps/PlatformBenchmarks/Program.cs [hunk 3]
```diff
@@ -3,19 +3,21 @@
...
 namespace PlatformBenchmarks
 {
     public class Program
     {
         public static void Main(string[] args)
         {
+            Console.WriteLine(BenchmarkApplication.ApplicationName);
+            Console.WriteLine(BenchmarkApplication.Paths.Plaintext);
+            Console.WriteLine(BenchmarkApplication.Paths.Json);
+            DateHeader.SyncDateTimer();
```
Comment (davidfowl) [inline benchmarkapps/PlatformBenchmarks/Program.cs]: Does this really make a difference?

Comment (benaadams) [reply benchmarkapps/PlatformBenchmarks/Program.cs]: Does an upfront initalize of the statics in DateHeader https://github.com/aspnet/KestrelHttpServer/pull/2569#issuecomment-388659044, so its not checked every call - normally wouldn't have much impact, but its being done 7 million times a second so...
```diff
 G_M36647_IG20:
-       48B94829EC53A1020000 mov      rcx, 0x2A153EC2948
-       BA02000000           mov      edx, 2
-       E8D561825F           call     CORINFO_HELP_GETSHARED_NONGCSTATIC_BASE
-       48B94829EA43F7010000 mov      rcx, 0x1F743EA2948
+       48B94829EC53A1020000 mov      rcx, 0x2A153EC2948
        488B09               mov      rcx, gword ptr [rcx]
        4885C9               test     rcx, rcx
        7506                 jne      SHORT G_M36647_IG21
        33D2                 xor      rdx, rdx
        33C0                 xor      eax, eax
        EB07                 jmp      SHORT G_M36647_IG22
```

**Comment by benaadams** [comment]
@sebastienros what's the process to update this in [TechEmpower/FrameworkBenchmarks/.../CSharp/aspnetcore/PlatformBenchmarks](https://github.com/TechEmpower/FrameworkBenchmarks/tree/master/frameworks/CSharp/aspnetcore/PlatformBenchmarks)?

**Comment by sebastienros** [comment]
I will simply update the files with this new version in a PR. But first I will get updated results to be sure we are not regressing. There is always a risk. Tomorrow morning. If the dashboard ones are ok then I won't hesitate much.

**Pull request merged**